scheds/rust/scx_lavd/src/bpf/main.bpf.c

```diff
@@ -0,0 +1,1749 @@
+}
+
+static bool slice_fully_consumed(struct cpu_ctx *cpuc, struct task_ctx *taskc)
+{
+	u64 run_time_ns = 0;
+
+	if (taskc->last_stop_clk > taskc->last_start_clk)
+		run_time_ns = taskc->last_stop_clk - taskc->last_start_clk;
+
+	return run_time_ns >= taskc->slice_ns;
+}
+
+static void adjust_slice_boost(struct cpu_ctx *cpuc, struct task_ctx *taskc)
+{
+	/*
+	 * Count how many times a task completely consumed the assigned time
+	 * slice to boost the task's slice when CPU is under-utilized. If not
+	 * fully consumed, decrease the slice boost priority by half.
+	 */
+	if (slice_fully_consumed(cpuc, taskc)) {
+		if (taskc->slice_boost_prio < LAVD_SLICE_BOOST_MAX)
+			taskc->slice_boost_prio++;
+	}
+	else {
+		if (taskc->slice_boost_prio)
+			taskc->slice_boost_prio >>= 1;
+	}
+}
+
+void BPF_STRUCT_OPS(lavd_stopping, struct task_struct *p, bool runnable)
+{
+	struct cpu_ctx *cpuc;
+	struct task_ctx *taskc;
+
+	/*
+	 * Reduce the task load.
+	 */
+	cpuc = get_cpu_ctx();
+	if (!cpuc) {
+		scx_bpf_error("cpu_ctx_stor lookup failed");
+		return;
+	}
+
+	taskc = bpf_task_storage_get(&task_ctx_stor, p, 0, 0);
+	if (!taskc) {
+		scx_bpf_error("task_ctx_stor lookup failed");
+		return;
+	}
+
+	if (transit_task_stat(taskc, LAVD_TASK_STAT_STOPPING))
+		update_stat_for_stop(p, taskc, cpuc);
+
+	/*
+	 * Adjust slice boost for the task's next schedule.
+	 */
+	adjust_slice_boost(cpuc, taskc);
+}
+
+void BPF_STRUCT_OPS(lavd_quiescent, struct task_struct *p, u64 deq_flags)
+{
+	struct task_ctx *taskc;
+	u64 now, interval;
+
+	/*
+	 * If a task @p is dequeued from a run queue for some other reason
+	 * other than going to sleep, it is an implementation-level side
+	 * effect. Hence, we don't care this spurious dequeue.
+	 */
+	if (!(deq_flags & SCX_DEQ_SLEEP))
+		return;
+
+	/*
+	 * When a task @p goes to sleep, its associated wait_freq is updated.
+	 */
+	taskc = bpf_task_storage_get(&task_ctx_stor, p, 0, 0);
+	if (!taskc) {
+		scx_bpf_error("task_ctx_stor lookup failed");
+		return;
+	}
+
+	now = bpf_ktime_get_ns();
+	interval = now - taskc->last_wait_clk;
+	taskc->wait_freq = calc_avg_freq(taskc->wait_freq, interval);
+	taskc->last_wait_clk = now;
+}
+
+void BPF_STRUCT_OPS(lavd_cpu_online, s32 cpu)
+{
+	/*
+	 * When a cpu becomes online, reset its cpu context and trigger the
+	 * recalculation of the global cpu load.
+	 */
+	struct cpu_ctx *cpuc = get_cpu_ctx_id(cpu);
+
+	if (!cpuc) {
+		scx_bpf_error("cpu_ctx_stor lookup failed");
+		return;
+	}
+
+	memset(cpuc, 0, sizeof(*cpuc));
+
+	__sync_fetch_and_add(&nr_cpus_onln, 1);
+	update_sys_cpu_load();
+}
+
+void BPF_STRUCT_OPS(lavd_cpu_offline, s32 cpu)
+{
+	/*
+	 * When a cpu becomes offline, trigger the recalculation of the global
+	 * cpu load.
+	 */
+	struct cpu_ctx *cpuc = get_cpu_ctx_id(cpu);
+
+	if (!cpuc) {
+		scx_bpf_error("cpu_ctx_stor lookup failed");
+		return;
+	}
+
+	memset(cpuc, 0, sizeof(*cpuc));
+
+	__sync_fetch_and_sub(&nr_cpus_onln, 1);
+	update_sys_cpu_load();
+}
```

Comment:
Out of curiosity, does steamdeck use CPU on/offline for power management?

Reply:
No, but the current SteamDeck firmware reports that 2x more CPUs are available so I used the number of online CPUs.